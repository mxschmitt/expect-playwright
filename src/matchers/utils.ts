import type { Page, ElementHandle, Frame } from "playwright-core"
import { PageWaitForSelectorOptions } from "../../global"

export type ExpectInputType = Page | Frame | ElementHandle

const isElementHandle = (value: ExpectInputType): value is ElementHandle => {
  return value.constructor.name === "ElementHandle"
}

export const getFrame = async (value: ExpectInputType) => {
  return isElementHandle(value) ? value.contentFrame() : value
}

const isObject = (value: unknown) =>
  typeof value === "object" && !(value instanceof RegExp)

export type InputArguments = [
  Page | ElementHandle,
  string?,
  (string | PageWaitForSelectorOptions)?,
  PageWaitForSelectorOptions?
]

export const getElementHandle = async (
  args: InputArguments,
  valueArgCount = 1
) => {
  // Pluck the options off the end first
  const options =
    args.length > 1 && isObject(args[args.length - 1])
      ? (args.pop() as PageWaitForSelectorOptions)
      : {}

  // Next, pluck the number of args required by the matcher (defaults to 1)
  const expectedValue = args.splice(-valueArgCount, valueArgCount) as string[]

  // Finally, we can find the element handle
  let elementHandle = (await getFrame(args[0])) ?? args[0]

  // If the user provided a page or iframe, we need to locate the provided
  // selector or the `body` element if none was provided.
  if (!isElementHandle(elementHandle)) {
    const selector = args[1] ?? "body"

    try {
      elementHandle = (await elementHandle.waitForSelector(selector, options))!
    } catch (err) {
      throw new Error(`Timeout exceed for element ${quote(selector)}`)
    }
  }

  return [elementHandle, expectedValue] as const
}

export const quote = (val: string | null) => (val === null ? "" : `'${val}'`)

export const getMessage = (
  { isNot, promise, utils, expand }: jest.MatcherContext,
  matcher: string,
  expected: unknown,
  received: unknown,
  expectedHint: string | undefined = undefined
) => {
  const message = isNot
    ? `Expected: not ${utils.printExpected(expected)}`
    : utils.printDiffOrStringify(
        expected,
        received,
        "Expected",
        "Received",
        expand
      )

  return (
    utils.matcherHint(matcher, undefined, expectedHint, { isNot, promise }) +
    "\n\n" +
    message
  )
}

export const compareText = (
  expectedValue: string | RegExp,
  actualValue: string | null
) => {
  return typeof expectedValue === "string"
    ? expectedValue === actualValue
    : expectedValue.test(actualValue ?? "")
}
