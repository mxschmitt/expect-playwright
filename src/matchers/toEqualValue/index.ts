import { SyncExpectationResult } from 'expect/build/types'
import { getElementText, quote, InputArguments } from '../utils'

const toEqualValue = async (...args: InputArguments): Promise<SyncExpectationResult> => {
  try {
    const { elementHandle, selector, expectedValue } = await getElementText(...args)
    /* istanbul ignore next */
    const actualTextContent = await elementHandle.evaluate((el) => (el as HTMLInputElement).value)
    if (actualTextContent?.includes(expectedValue)) {
      return {
        pass: true,
        message: () => `${quote(expectedValue)} is included in ${quote(actualTextContent)}.`
      }
    }
    return {
      pass: false,
      message: () => `${quote(expectedValue)} is not included in ${quote(actualTextContent)}${selector ? ' of ' + quote(selector) + "." : '.'}`
    }
  } catch (err) {
    return {
      pass: false,
      message: () => err.toString()
    }
  }
}

export default toEqualValue