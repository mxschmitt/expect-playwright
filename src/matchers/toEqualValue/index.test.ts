import { testWrapper } from "../tests/utils"

import toEqualValue from '.'

describe("toEqualValue", () => {
  afterEach(async () => {
    await page.evaluate(() => document.body.innerHTML = "")
  })
  describe("selector", () => {
    it("positive", async () => {
      await page.evaluate(() => {
        document.write(`<input id="foobar" value="bar"/>`)
      })
      const result = await toEqualValue(page, "#foobar", "bar")
      expect(testWrapper(result)).toBe(true)
      expect(result.message()).toMatchSnapshot()
    })
    it("negative", async () => {
      await page.evaluate(() => {
        document.write(`<input id="foobar" value="bar"/>`)
      })
      expect(testWrapper(await toEqualValue(page, "#foobar", "not-existing"))).toThrowErrorMatchingSnapshot()
    })
  })
  describe("element", () => {
    it("positive", async () => {
      await page.evaluate(() => {
        document.write(`<input id="foobar" value="bar"/>`)
      })
      const element = await page.$("#foobar")
      expect(element).not.toBe(null)
      expect(testWrapper(await toEqualValue(element!, "bar"))).toBe(true)
    })
    it("negative", async () => {
      await page.evaluate(() => {
        document.write(`<input id="foobar" value="bar"/>`)
      })
      const element = await page.$("#foobar")
      expect(element).not.toBe(null)
      expect(testWrapper(await toEqualValue(element!, "not-existing"))).toThrowErrorMatchingSnapshot()
    })
  })
  describe("timeout", () => {
    it("should throw an error after the timeout exceeds", async () => {
      const start = new Date().getTime()
      expect(testWrapper(await toEqualValue(page, "#foobar", "bar"))).toThrowErrorMatchingSnapshot()
      const duration = new Date().getTime() - start
      expect(duration).toBeLessThan(1500)
    })
  })
})