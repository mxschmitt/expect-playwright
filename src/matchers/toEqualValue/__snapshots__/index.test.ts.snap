// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`toEqualValue element negative 1`] = `"'not-existing' is not included in 'bar'."`;

exports[`toEqualValue selector negative 1`] = `"'not-existing' is not included in 'bar' of '#foobar'."`;

exports[`toEqualValue selector positive 1`] = `"'bar' is included in 'bar'."`;

exports[`toEqualValue timeout should throw an error after the timeout exceeds 1`] = `"Error: Timeout exceed for element '#foobar'"`;
